quiche/src/recovery/mod.rs

```diff
@@ -664,89 +778,80 @@ impl Recovery {
         // This will also trigger sending an ACK and retransmitting frames like
         // HANDSHAKE_DONE and MAX_DATA / MAX_STREAM_DATA as well, in addition
         // to CRYPTO and STREAM, if the original packet carried them.
-        for unacked in unacked_iter {
-            self.lost[epoch].extend_from_slice(&unacked.frames);
-        }
+        epoch.lost_frames.extend(unacked_frames.cloned());
+
+        self.pacer
+            .on_retransmission_timeout(!epoch.lost_frames.is_empty());
 
         self.set_loss_detection_timer(handshake_status, now);
 
         trace!("{} {:?}", trace_id, self);
-
         (0, 0)
     }
 
     pub fn on_pkt_num_space_discarded(
         &mut self, epoch: packet::Epoch, handshake_status: HandshakeStatus,
         now: Instant,
     ) {
-        let unacked_bytes = self.sent[epoch]
-            .iter()
-            .filter(|p| {
-                p.in_flight && p.time_acked.is_none() && p.time_lost.is_none()
-            })
-            .fold(0, |acc, p| acc + p.size);
-
-        self.bytes_in_flight = self.bytes_in_flight.saturating_sub(unacked_bytes);
-
-        self.sent[epoch].clear();
-        self.lost[epoch].clear();
-        self.acked[epoch].clear();
-
-        self.time_of_last_sent_ack_eliciting_pkt[epoch] = None;
-        self.loss_time[epoch] = None;
-        self.loss_probes[epoch] = 0;
-        self.in_flight_count[epoch] = 0;
-
+        let epoch = &mut self.epochs[epoch];
+        self.bytes_in_flight = self
+            .bytes_in_flight
+            .saturating_sub(epoch.discard(&mut self.pacer));
         self.set_loss_detection_timer(handshake_status, now);
     }
 
     pub fn on_path_change(
-        &mut self, epoch: packet::Epoch, now: Instant, trace_id: &str,
+        &mut self, epoch: packet::Epoch, now: Instant, _trace_id: &str,
     ) -> (usize, usize) {
-        self.detect_lost_packets(epoch, now, trace_id)
+        let (lost_bytes, lost_packets) =
+            self.detect_and_remove_lost_packets(epoch, now);
```

Comment:
I see no call to `self.pacer.on_congestion_event` after `detect_and_remove_lost_packets` in this branch.  Is this intentional?

Reply:
Yes, I think it is, path change is not a true congestion event, in fact maybe cc state needs to be reset completely in this case? But I would avoid big changes to working code, before it is merged, better to improve after